Update disconnected UpdateService operator documentation (#137)
openshift/cincinnati-operator: docs/disconnected-updateservice-operator.md

You might want to review the documentation around disconnected registries to learn about mirroring OpenShift Releases.

```
export UPSTREAM_REGISTRY='quay.io'
export PRODUCT_REPO='openshift-release-dev'
export RELEASE_NAME='ocp-release'
export OCP_RELEASE='4.5.2-x86_64'
export LOCAL_REGISTRY='my-disconnected-registry.example.com:5000'
export LOCAL_SECRET_JSON='/path/to/pull_secret.json'

oc adm -a ${LOCAL_SECRET_JSON} release mirror \
--from=${UPSTREAM_REGISTRY}/${PRODUCT_REPO}/${RELEASE_NAME}:${OCP_RELEASE} \
--to=${LOCAL_REGISTRY}/ocp4 \
--to-release-image=${LOCAL_REGISTRY}/ocp4/release:${OCP_RELEASE}

export OCP_RELEASE='4.5.3-x86_64'

oc adm -a ${LOCAL_SECRET_JSON} release mirror \
--from=${UPSTREAM_REGISTRY}/${PRODUCT_REPO}/${RELEASE_NAME}:${OCP_RELEASE} \
--to=${LOCAL_REGISTRY}/ocp4 \
--to-release-image=${LOCAL_REGISTRY}/ocp4/release:${OCP_RELEASE}
```

## Deploy UpdateService instance

```
cat create-updateservice-instance.yaml
---
apiVersion: updateservice.operator.openshift.io/v1
kind: UpdateService
metadata:
  name: karampok
  namespace: openshift-update-service
spec:
  replicas: 1
  releases: "{DISCONNECTED_REGISTRY}/ocp4/release"

>   #releases: "{DISCONNECTED_REGISTRY}/ocp4 // OCP Release Image and Release Content in the same path

Reviewer: Lets remove this, we should never recommend mixing release images and content in one repo, as it would most certainly make image scraping complicated

  graphDataImage: "{DISCONNECTED_REGISTRY}/updateservice/graph-data:v4"
```
    
### Known Issues

#### OOMKilled

If you have you mirrored your OpenShift Release Image and Release Content
with an identical path, UpdateService's Graph Builder will take more memory
than allowed by the Deployment configuration and the pod will be OOMKilled.
If this is your case you can copy the release image to a different namespace
in your registry.

```
skopeo copy docker://${DISCONNECTED_REGISTRY}/ocp4:4.5.2-x86_64 docker://${DISCONNECTED_REGISTRY}/release:4.5.2-x86_64 --authfile=/path/to/pull_secret.json
skopeo copy docker://${DISCONNECTED_REGISTRY}/ocp5:4.5.3-x86_64 docker://${DISCONNECTED_REGISTRY}/release:4.5.3-x86_64 --authfile=/path/to/pull_secret.json
```

#### CreateRouteFailed

    

> If you see `ReconcileCompleted` status as `false` with reason `CreateRouteFailed` caused by `host must conform to DNS 1123 naming convention`

Reviewer: Is it worth mentioning? It looks like a fairly common error

Reviewer 2: This depends on the base domain of the cluster, it's a limitation on the route names length. I believe keeping it here doesn't hurt, but up to you.

Reviewer: Sounds like its a common problem, so lets keep it - it needs a link to bugzilla though

Reviewer: If operator allows them to configure settings which end up in an invalid route I'd say its a bug

Author: I do not think there is bug here, if I understand it is just a small reminder that 
basedomain + object name is the DNS name which should feed the length.

and `must be no more than 63 characters`, try creating the Update Service with a shorter name.
    
## Make client cluster use UpdateService instance

> Once UpdateService starts, there is a public URL (OCP route) available. This URL is should be configured in the client cluster. 

Reviewer: Needs a link to "update ClusterVersion's url" doc section

```
oc adm upgrade
>  Cluster version is 4.8.22

>  Upgradeable=False
>    Reason: AdminAckRequired
>    Message: Kubernetes 1.22 and therefore OpenShift 4.9 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6329921 for details and instructions.
>  warning: Cannot display available updates:
>    Reason: RemoteFailed
>    Message: Unable to retrieve available updates: Get "https://api.openshift.com/api/upgrades_info/v1/graph?arch=amd64&channel=stable-4.8&id=6f878e7c-cb81-41de-a954-bb6766548318&version=4.8.22": dial tcp 18.210.64.188:443: connect: network is unreachable
# ...  old public update service url ...

POLICY_ENGINE_GRAPH_URI="$(oc -n openshift-update-service get updateservice karampok -o jsonpath='{.status.policyEngineURI}/api/upgrades_info/v1/graph')"
PATCH="{\"spec\":{\"upstream\":\"${POLICY_ENGINE_GRAPH_URI}\"}}"
oc patch clusterversion version -p $PATCH --type merge
```

### Make client cluster trust UpdateService server certificate

> Unless your UpdateService tls server certificate is somehow trusted by the client OCP cluster (e.g. well known root CA), then you probably get the

Reviewer: Needs a link to addtionalTrustedCA doc section from proxy server

> Cluster version is 4.8.22
> Upgradeable=False
>   Reason: AdminAckRequired
>   Message: Kubernetes 1.22 and therefore OpenShift 4.9 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6329921 for details and instructions.
> warning: Cannot display available updates:
>   Reason: RemoteFailed
>   Message: Unable to retrieve available updates: Get "https://karampok-policy-engine-route-openshift-update-service.apps.bcn.hub-virtual.lab/api/upgrades_info/v1/graph?arch=amd64&channel=stable-4.8&id=6f878e7c-cb81-41de-a954-bb6766548318&version=4.8.22": 
#  ... x509: certificate signed by unknown authority ....
```

If the client cluster, is different than the cluster where UpdateService was installed, you can either add the CA cert 
on the node or in the proxy configuration: 

```
    oc create configmap trusted-updateservice-ca -n openshift-config --from-file=ca-bundle.crt=OCPRootCA.pem
    oc patch proxy/cluster -p '{"spec":{"trustedCA":{"name":"trusted-updateservice-ca"}}}' --type merge
    oc edit proxy/cluster
    # ... wait until pods are restarting ..
```

Finally you should get the following:

```

> dan adm upgrade

Reviewer: ```suggestion
oc adm upgrade
```

> Cluster version is 4.8.22
> Upgradeable=False
>  Reason: AdminAckRequired
>  Message: Kubernetes 1.22 and therefore OpenShift 4.9 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6329921 for details and instructions.
> Updates:
> VERSION IMAGE
> 4.8.23  {DISCONNECTED_REGISTRY}/ocp4/release@sha256:3fab205d36c66825423274eac90f4c142a18cdf358b4a666a1783d325afba860
```

## Update Client Cluster

> We can force the update using the CLI

Reviewer: Lets not mention `force` as the cluster may lose supported status. Needs to be reformatted into `in case of emergency` or smth similar. 
Same for "--allow-upgrade-with-warnings".

"--to-latest" is also probably not a great idea, lets rework into "to desired version" and adding that "--to-latest" is a shortcut for that

```
oc adm upgrade --to-latest --allow-upgrade-with-warnings --force
```
    

> ### Add release signature

Reviewer: ```suggestion
### Add release signature verification config
```

> To make the release safe, so no extra flags are needed, we can add a release signature configmap:

Reviewer: This needs to be moved before "start the upgrade". Signature verification is important to prevent supply chain attacks

> Updates:
> VERSION IMAGE
> 4.8.23  {DISCONNECTED_REGISTRY}/ocp4/release@sha256:3fab205d36c66825423274eac90f4c142a18cdf358b4a666a1783d325afba860
```

### Add release signature verification config

To make the release safe, so no extra flags are needed, we can add a release signature configmap:

```
export release=4.8.23
export digest=$(oc adm release info ${DISCONNECTED_REGISTRY}/ocp4/release:${release}-x86_64 --registry-config=${XDG_RUNTIME_DIR}/containers/auth.json -o json | jq  -r .digest)
export signature=$(curl -s "https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/${digest//:/=}/signature-1" | base64 -w0 && echo)
cat <<EOF | oc create -f - 
apiVersion: v1
kind: ConfigMap
metadata:
  name: release-image-${release}
  namespace: openshift-config-managed
  labels:
    release.openshift.io/verification-signatures: ""
binaryData:
  ${digest//:/-}: ${signature}"
EOF
```

## Update Client Cluster

We can start the update using the CLI and executing:

```
oc adm upgrade --to=<version>
```
    

> In case of emergency where the update is being blocked, the argument `--allow-upgrade-with-warnings` can be used.

Reviewer: ```suggestion
If this command throws any warnings which are not important for cluster upgrade, use the argument `--allow-upgrade-with-warnings`. If user wants to upgrade the cluster to a cluster which is not available in the graph, use `--force` argument. Note, that cluster upgraded using any of these arguments may be in unsupportable state
```
